habitat-hitl/habitat_hitl/environment/controllers/llm_controller.py

```diff
@@ -0,0 +1,167 @@
+    SingleAgentBaselinesController,
+)
+
+
+class LLMController(SingleAgentBaselinesController):
+    """Controller for single LLM controlled agent."""
+
+    def __init__(
+        self,
+        agent_idx: int,
+        is_multi_agent: bool,
+        config: DictConfig,
+        gym_habitat_env: GymHabitatEnv,
+    ):
+        self._config = config
+        self._is_multi_agent = is_multi_agent
+        self._gym_habitat_env = gym_habitat_env
+        self._habitat_env = gym_habitat_env.unwrapped.habitat_env
+        self._agent_idx = agent_idx
+        # TODO: gather this from config
+        self._agent_action_length = 28
+        self._thread: Union[None, threading.Thread] = None
+        self._low_level_actions: Union[None, dict, np.ndarray] = {}
+        self._iter = 0
+        self._skip_iters = 0
+        logging.basicConfig(
+            filename="/home/priyamp/hitl/act_timing.log",
+            filemode="a",
+            format="%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s",
+            datefmt="%H:%M:%S",
+            level=logging.DEBUG,
+            force=True,
+        )
+        self._analysis_logger = logging.getLogger("LLMController")
+        self._analysis_logger.debug("LLMController initialized")
+
+        with habitat.config.read_write(self._config):
+            fix_config(self._config)
+        seed = 47668090
+        with habitat.config.read_write(self._config):
+            self._config = setup_config(self._config, seed)
+        self.planner: LLMPlanner = None
+        self.environment_interface: EnvironmentInterface = None
+
+        # NOTE: this is creating just one agent. Habitat-LLM has code for creating
+        # multiple processes/agents in one go. I am only prototyping single process, as
+        # I assume the onus of creating multiple processes is on the user/hitl_driver,
+        # and this code will be called once per Sim instantiation
+        self.initialize_environment_interface()
+        self.initialize_planner()
+        self.info: Dict[str, Any] = {}
+
+    def initialize_planner(self):
+        # NOTE: using instantiate here, but given this is planning for a single agent
+        # always will this ever be an option of Centralized vs Decentralized? Maybe
+        # DAG...?
+        # NOTE: assuming use of DecentralizedLLMPlanner here
+        planner = instantiate(self._config.evaluation.agents.agent_0.planner)
+        self.planner = planner(env_interface=self.environment_interface)
+        self.planner.agents = self.initialize_agents(
+            self._config.evaluation.agents
+        )
+
+    def initialize_agents(self, agent_configs):
+        agents = []
+        for _, agent_conf in agent_configs.items():
+            # Instantiate the agent
+            agent = Agent(
+                agent_conf.uid, agent_conf.config, self.environment_interface
+            )
+
+            # Make sure that its unique by adding to the set
+            agents.append(agent)
```

Comment:
> Make sure that its unique by adding to the set

The container is not a set. Typo?

Reply:
Good catch. This is cargo-cult artifact.